### PR TITLE
Functional support mma.sync + ldmatrix for F16 datatype 

### PR DESCRIPTION
**Pull request by manishucsd**
Adds native Tensor Core (F16) support for NVIDIA GPUs. Specifically, adds support to target `mma.sync.aligned.m16n8k16.row.col.f16.f16.f16.f16` and `ldmatrix` instructions on NVIDIA A100 GPUs. For now, these instructions are not enabled by default and requires using (`--iree-codegen-llvmgpu-use-mma-sync=true`). 

IREE runs row-row matmul and the native Tensor Core support requires `ldmatrix` for operandA and `ldmatrix.trans` for operandB. `ldmatrix.trans` transposes the data while loading to use Tensor Core `mma*row.col*` instruction. 

There are LLVM/MLIR upstream pathes and previous IREE PRs that enable targeting native Tensor Core F16 support.

A few LLVM/MLIR upstream patches:
- [[mlir][NVGPU] Adding Support for cp_async_zfill via Inline Asm](https://reviews.llvm.org/D132269)
- [[mlir][NVGPU] Handle Native mma.sync sizes and ldmatrix(x4) for matrixB](https://reviews.llvm.org/D135749)
- [[mlir][NVGPU] Fix affine maps computing indices for LdMatrixOp srcMemref](https://reviews.llvm.org/D138978) 
- [[mlir][NVGPU] Fix NVGPU Bazel Layering](https://reviews.llvm.org/D141982)

Previous IREE PRs:
- [Support GEMM Pipelining *without* Epilogue Peeling](https://github.com/iree-org/iree/pull/10388)
- [Add support for GEMM e2e Test For CUDA backend on F16 input](https://github.com/iree-org/iree/pull/10842)

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp [hunk 1]
```diff
@@ -101,7 +101,7 @@ static void getTensorCoreConfig(
   // Tile sizes are skewed towards small matmul for now. Long term the plan is
   // to not rely on hardcoded configurations.
   if (isFp16) {
-    tileSizes.push_back(TileWorkgroupSizePair({{32, 32, 32}, {64, 2, 1}}));
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:104]: I don't think we want to change tile size in this PR. If we need this please make it a separate PR so that we could easily figure out if this causes changes and keep the older configuration as a fall back.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp [hunk 2]
```diff
@@ -119,10 +120,132 @@ static Optional<SmallVector<int64_t>> getGPUTCNativeVectorSize(Operation *op) {
   return std::nullopt;
 }
 
+static int getVectorContractOpOperandId(
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp:123]: make it an `Optional<int>`

Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp:123]: also please add doc

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp [hunk 3]
```diff
@@ -119,10 +120,132 @@ static Optional<SmallVector<int64_t>> getGPUTCNativeVectorSize(Operation *op) {
...
+  vector::ContractionOp contractOp, Operation *op) {
+  if (contractOp.getLhs() == op->getResult(0)) return 0;
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp:125]: It looks odd to do op->getResult(0), you should probably pass directly the `Value` to the function

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp [hunk 4]
```diff
@@ -119,10 +120,132 @@ static Optional<SmallVector<int64_t>> getGPUTCNativeVectorSize(Operation *op) {
...
+  if (contractOp.getRhs() == op->getResult(0)) return 1;
+  if (contractOp.getAcc() == op->getResult(0)) return 2;
+  return -1;
+}
+
+// Returns operand id for vector::ContractOp where the vector::TransferReadOp
+// is consumed. 
+static int getVectorContractOpOperandIdForVectorReadOp(Operation *op) {
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp:133]: ditto make it `Optional<int>`

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp [hunk 5]
```diff
@@ -119,10 +120,132 @@ static Optional<SmallVector<int64_t>> getGPUTCNativeVectorSize(Operation *op) {
...
 
+static int getVectorContractOpOperandId(
+  vector::ContractionOp contractOp, Operation *op) {
+  if (contractOp.getLhs() == op->getResult(0)) return 0;
+  if (contractOp.getRhs() == op->getResult(0)) return 1;
+  if (contractOp.getAcc() == op->getResult(0)) return 2;
+  return -1;
+}
+
+// Returns operand id for vector::ContractOp where the vector::TransferReadOp
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp:131]: nit: use `///` for function doc

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp [hunk 6]
```diff
@@ -119,10 +120,132 @@ static Optional<SmallVector<int64_t>> getGPUTCNativeVectorSize(Operation *op) {
...
+    Operation *op) {
+  
+  // Shape of native Tensor Core GPU mma.sync operations.
+  int64_t mmaShapeM = 16;
+  int64_t mmaShapeN = 8;
+  int64_t mmaShapeK;
+
+  // Shape the mma.sync warp-level operation.
+  if (auto contract = dyn_cast<vector::ContractionOp>(op)) {
+    auto sourceType = contract.getLhsType().getElementType();
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp:156]: spell out type here, following this convention:
https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp [hunk 7]
```diff
@@ -119,10 +120,132 @@ static Optional<SmallVector<int64_t>> getGPUTCNativeVectorSize(Operation *op) {
...
+    if (resultElementType.isF32()) {
+      // TODO
+      op->emitError() << "TODO Implement shape for F32 loads in getMmaNativeVectorSize";
+      return std::nullopt;
+    }
+  }
+  return std::nullopt;
+}
+
+static Optional<SmallVector<int64_t>> getGPUTensorCoreNativeVectorSize(
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp:237]: can we add a test for this logic?

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTileAndDistribute.cpp [hunk 8]
```diff
@@ -235,6 +236,7 @@ struct LLVMGPUTileAndDistributePass
       insertBarriersAroundSharedMemoryCopy(funcOp);
     }
 
+
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTileAndDistribute.cpp:239]: remove extra line

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTileAndDistribute.cpp [hunk 9]
```diff
@@ -221,6 +221,7 @@ struct LLVMGPUTileAndDistributePass
 
     int64_t flatWorkgroupSize =
         workgroupSize[0] * workgroupSize[1] * workgroupSize[2];
+
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTileAndDistribute.cpp:224]: remove extra line

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorToGPU.cpp [hunk 10]
```diff
@@ -137,19 +137,12 @@ struct LLVMGPUVectorToGPUPass
                                             std::move(flatternpatterns)))) {
       return signalPassFailure();
     }
-    RewritePatternSet patterns(funcOp.getContext());
-    mlir::vector::populateCastAwayVectorLeadingOneDimPatterns(patterns);
-    populatePrepareVectorToMMAPatterns(patterns, llvmgpuUseMMASync);
-    if (failed(applyPatternsAndFoldGreedily(getOperation(),
-                                            std::move(patterns)))) {
-      return signalPassFailure();
-    }
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorToGPU.cpp:146]: any reason to remove those patterns here? Sounds like we apply `populateCastAwayVectorLeadingOneDimPatterns` later. I would leave it as is

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp [hunk 11]
```diff
@@ -218,6 +218,7 @@ void addGPUMatmulTensorCorePassPipeline(OpPassManager &pm,
       createLLVMGPUTileAndDistribute(/*distributeToWarp=*/true));
   nestedModulePM.addNestedPass<func::FuncOp>(
       createRemoveSingleIterationLoopPass());
+
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp:221]: remove

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp [hunk 12]
```diff
@@ -253,10 +254,11 @@ void addGPUMatmulTensorCorePassPipeline(OpPassManager &pm,
   // Vector -> MMA ops
   nestedModulePM.addNestedPass<func::FuncOp>(
       memref::createFoldMemRefAliasOpsPass());
+  nestedModulePM.addPass(createCanonicalizerPass());
+  nestedModulePM.addPass(createCSEPass());
   nestedModulePM.addNestedPass<func::FuncOp>(createLLVMGPUVectorToGPU());
   nestedModulePM.addPass(createCanonicalizerPass());
   nestedModulePM.addPass(createCSEPass());
-
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp:259]: remove

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp [hunk 13]
```diff
@@ -94,6 +94,7 @@ LogicalResult verifyGPUMatmulTensorCorePipeline(
     Operation *op, IREE::Codegen::LoweringConfigAttr loweringConfig,
     IREE::Codegen::TranslationInfoAttr translationInfo,
     ArrayRef<int64_t> workgroupSize) {
+      return success();
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp:97]: remove

File: tests/e2e/matmul/generate_e2e_matmul_tests.py [hunk 14]
```diff
@@ -137,7 +137,7 @@ def get_test_shapes(shapes_id: ShapesId):
         # (see get_test_generators).
     ]
   if shapes_id == ShapesId.GPU_LARGE:
-    return [TestShape(m=256, k=128, n=512)]
```
Comment (ThomasRaoux) [inline tests/e2e/matmul/generate_e2e_matmul_tests.py:140]: why do we need this change?

**Comment by MaheshRavishankar** [comment]
Drive by: Please add some description (people do see and appreciate PR hygeine)

File: tests/e2e/matmul/generate_e2e_matmul_tests.py (hunk 14 above)
Comment (manishucsd) [reply tests/e2e/matmul/generate_e2e_matmul_tests.py]: I have all my notes and debugging pictures for 512x256x128. It is fine to revert it to the original, but you might see the m and n flipped again in some time the next PR :)

**Comment by manishucsd** [comment]
> Drive by: Please add some description (people do see and appreciate PR hygeine)

Thank you for bringing this to my attention. I was hoping to start this PR and start CI pipelines, clean it, and add the description as I finalize the PR by fixing build issues and incorporating review comments. Not fully ready to merge yet, but added the description. I will add the description as the first thing from the next time :).

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp (hunk 7 above)
Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp:238]: If it is ok with you, let us handle it in the next PR on F32 Tensor Core support via `mma.sync`? Along with enabling F32 lit tests in `LLVMGPU/test/nvvm_pipeline_test.mlir`

**Review by ThomasRaoux — APPROVED** [review]
Looks good!
